[RFC/RFT] iio: dac: ad5766: add reading and writing

denisgabriel5 opened the pull request:
I may be missing something, but I don't know what. 
When I try to test it with an oscilloscope, it doesn't work.
This is still a work in progress.

Signed-off-by: Denis Gheorghescu <denis.gheorghescu@analog.com>

commodo commented on drivers/iio/dac/ad5766.h:
```diff
@@ -0,0 +1,89 @@
+#include <linux/mutex.h>
+#include <linux/kernel.h>
+
+#define AD5766_CMD_NOP_MUX_OUT			0x00
```
you can remove the entire header code in the C-file;
usually, when headers are used, code is shared between more than 1 C file;
in this case, the driver is standalone unit and won't need to share any of it's definitions with other drivers.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -7,53 +7,254 @@
+	indio_dev->num_channels = st->chip_info->num_channels;
 
-	return devm_iio_device_register(&spi->dev, indio_dev);
+	return iio_device_register(indio_dev);
```
why `devm_iio_device_register()` -> `iio_device_register()` ?
it doesn't look like it's needed;
also, when switching to  `iio_device_register()` , you need to call an equivalent ` `iio_device_unregister()`  on the `ad5766_remove` hook; meaning that you need to add a `ad5766_remove` hook

denisgabriel replied:
honestly, I forgot to switch back to `devm_iio_device_register`, I was trying to see something.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -7,53 +7,254 @@
+
+	mutex_lock(&indio_dev->mlock);
+	ret = _ad5766_spi_read(st, dac, val);
+	mutex_unlock(&indio_dev->mlock);
```
using `&indio_dev->mlock` in IIO drivers is discouraged; 
it was used in many old drivers, and there is an effort to clean drivers up;
but you still find some old ones that still use it, so if you reference stuff from them, you get here;

also, i'm not sure that you need a lock yet;
but if you are referencing ad5761 here and that has a lock, we can keep it until more stuff is added;

commodo replied:
i forgot to mention;
the alternative to using `indio_dev->mlock` is to define a lock on the driver's state struct;
you do need to init it on probe;
check for pattern `mutex_lock(&st->lock)`

commodo replied:
my bad; i think you need a lock;

denisgabriel replied:
yep, I tried to apply the same pattern from `ad5761`, but I've seen it in `ad5764` and `5686`.

so, should I keep it?

commodo replied:
maybe convert it to `st->lock`

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -7,53 +7,254 @@
+
+static int ad5766_read(struct iio_dev *indio_dev,
+			   enum ad5766_dac dac,
+			   __u16 *val)
```
__u16 -> u16 ?

denisgabriel replied:
it may be my editor's fault, but it says it doesn't recognise `u16`

nunojsa replied:
Most likely editor's fault :smile: . It should be defined in `linux/types.h` if I'm not mistaken...

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -7,53 +7,254 @@
 #include <linux/spi/spi.h>
+#include <linux/slab.h>
+#include <linux/sysfs.h>
+#include <linux/regulator/consumer.h>
```
maybe don't add all headers yet;
only the minimum you need;
i think `linux/regulator/consumer.h` is not needed yet

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -6,54 +6,327 @@
+	indio_dev->num_channels = st->chip_info->num_channels;
 
-	return devm_iio_device_register(&spi->dev, indio_dev);
+	return devm_iio_device_register(&spi_dev->dev, indio_dev);
```
why rename `spi` -> `spi_dev` ?
it's just patch noise;
maybe let's keep the name `spi`;

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -6,54 +6,327 @@
-	spi_set_drvdata(spi, indio_dev);
+	spi_set_drvdata(spi_dev, indio_dev);
 
-	st->spi = spi;
```
this moving around looks weird;
is this patch re-sorting stuff?

commodo replied:
hmm; it may be due to the rename of `spi` -> `spi_dev`
i'd remove it; it makes the patch too noisy

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -44,14 +319,14 @@ static int ad5766_probe(struct spi_device *spi)
+	indio_dev->info = &ad5766_info;
 	indio_dev->modes = INDIO_DIRECT_MODE;
-
-	ret = ad5766_setup(st);
```
curios: are you sure `ad5766_setup()` won't be needed?

denisgabriel replied:
well, I'll add it in the future if it's going to be needed

dbogdan replied:
This change is unrelated to the subject of this commit.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -6,35 +6,310 @@
-	int ret;
+
+	/// TODO: remove all prints below
+	dev_info(&spi->dev, "We're in probe!\n");
```
for the final code [when that happens] be sure to remember to remove these prints;

denisgabriel replied:
of course

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -44,14 +319,14 @@ static int ad5766_probe(struct spi_device *spi)
+	st->chip_info = &ad5766_chip_infos[type];
+
 	indio_dev->dev.parent = &spi->dev;
-	indio_dev->dev.of_node = spi->dev.of_node;
```
curios: why remove `indio_dev->dev.of_node = spi->dev.of_node;` ?

denisgabriel replied:
well, I looked at a few others drivers and there was no `indio_dev->dev.of_node = spi->dev.of_node;`

denisgabriel replied:
so I thought that I should leave in it only the things I understand (or at least the things that I have a vague idea of what they're doing)

dbogdan replied:
This change doesn't have anything to do with the reading and writing (the subject of this commit) - so please don't do it now.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -6,35 +6,310 @@
+
+	mutex_lock(&st->lock);
+	ret = _ad5766_spi_write(st, AD5766_CMD_WR_DAC_REG(dac), data);
+	mutex_unlock(&st->lock);
```
there should be a `mutex_init(&st->lock);` in probe;
there does not need to be a `mutex_uninit(&st->lock);` in remove; since the mutex data is static on the state struct;

denisgabriel commented:
changelog v3->v4

- added the initialization for mutex in function `probe`.

denisgabriel commented:
changelog v4->v5

denisgabriel commented:
changelog v5->v6

- remove useless define `AD5766_FULL_RESET_CODE`.

denisgabriel5 commented:
changelog v6->v7

- removed `be32_to_cpu` conversion & used only the second word;
- removed useless use of macro `AD5766_CMD_WR_DAC_REG`;
- code cleanup: aligned some `\` & some macro's values.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -6,35 +6,312 @@
+	st->data[0].d32 = cpu_to_be32(AD5766_CMD_READBACK_REG(dac));
+	st->data[1].d32 = cpu_to_be32(AD5766_CMD_NOP_MUX_OUT);
+
+	ret = spi_sync_transfer(st->spi, xfers, ARRAY_SIZE(xfers));
```
there should be an error check right after this SPI transaction;
```
if (ret)
    return  ret;
```
or something like that;
`*val ` shoudl not be assigned if an error occurred;

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -246,20 +242,20 @@ static int _ad5766_spi_read(struct ad5766_state *st,
 			.len = 3,
 			.cs_change = 1,
 		}, {
-			.tx_buf = &st->data[1].d8[1],
```
maybe squash this patch into the first;
no need to have a fix commit;

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -259,7 +261,7 @@ static int _ad5766_spi_read(struct ad5766_state *st,
 	ret = spi_sync_transfer(st->spi, xfers, ARRAY_SIZE(xfers));
 
-	*val = be32_to_cpu(st->data[2].d32);
+	*val = st->data[2].d16[1];
```
squash this into the first one as well;

nunojsa replied:
Same as the previous comment. `d16` is a `__be16` type. You should not mix it with "normal" cpu types. Here, you would need something like `*val = be16_to_cpu(st->data[2].d16[1])`.

commodo commented:
so, generally, i'd squash all 3 into a single patch;
the 2nd 2 patches are fix commits;

denisgabriel5 commented:
changelog v6->v7

- add error check.

denisgabriel5 commented:
changelog v8->v9

- squash commits.

nunojsa commented on drivers/iio/dac/ad5766.c:
```diff
@@ -6,35 +6,307 @@
+
+	st->data[0].d32 = AD5766_CMD_READBACK_REG(dac);
+	st->data[1].d32 = AD5766_CMD_NOP_MUX_OUT;
+
```
Sparse (compile your driver with `C=2` to see sparse complains...) will complain about this. Make sure you really want to use the restricted types __be32 (or be32). If so, you need to do something like `st->data[0].d32 = cpu_to_be32(AD5766_CMD_READBACK_REG(dac))`. The same on the other line... Note that your data will be swapped to BE if in LE machine...

denisgabriel5 replied:
> Most likely editor's fault. It should be defined in `linux/types.h` if I'm not mistaken...

Can you please be a little more explicit here? I'm not sure I understood what you meant by that comment.

nunojsa replied:
Hmm not sure how we got here ehehe. This was regarding this comment from @commodo 

> __u16 -> u16 ?

That you replied with

> it may be my editor's fault, but it says it doesn't recognise u16

denisgabriel5 replied:
Nevermind, GitHub scrambled my messages. Well, I tried including it and it still looks like it does not recognize it. I can push it like that if you guys are ok with it.

denisgabriel5 commented:
changelog v9->v10

- ditch `__beXX` type in favour of `uXX` type;
- renamed `ad5766_state->data`'s fields according to their type: 
`b8`-`byte`, `w16`-`word` & `d32`-`doubleword`;
- code cleanup.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -6,35 +6,308 @@
+		return -EINVAL;
+	}
+
+	return ad5766_write(indio_dev, chan->address, (__u16)val);
```
remove this ` (__u16)` cast

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -6,35 +6,308 @@
+
+	switch (m) {
+	case IIO_CHAN_INFO_RAW:
+		ret = ad5766_read(indio_dev, chan->address, (__u16 *)val);
```
remove this ` (__u16 9)` cast

commodo replied:
if you get issues about pointer compatibility when removing the cast, then it's a good idea to change the type of the argument from `u16 *` -> `int *` and propagate it downwards;

the idea is that [where possible] explicit pointer-casts are discouraged;

denisgabriel5 commented:
changelog v10->v11

- add `scale` & `offset` feature;
- change a few var types to remove some unnecessary casts;
- remove a couple of unnecessary features at the moment like 
`CALIBSCALE` & `CALIBBIAS`;
- rename a var for better readability in `_ad5766_spi_write()`;
- a bit of code cleanup.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -46,9 +45,7 @@
 	.channel = (_chan),										\
 	.address = (_chan),										\
 	.info_mask_separate = BIT(IIO_CHAN_INFO_RAW) |			\
-		BIT(IIO_CHAN_INFO_SCALE) |							\
```
why are these being removed?

denisgabriel5 commented:
changelog v11->v12

- remove off-topic changes like: `scale` & `offset` features and cleanup;
- rename a var for better readability in _ad5766_spi_write();
- a bit of code cleanup.

dbogdan commented on drivers/iio/dac/ad5766.c:
```diff
@@ -6,35 +6,308 @@
 #include <linux/spi/spi.h>
+#include <linux/slab.h>
+#include <linux/sysfs.h>
+#include <linux/types.h>
```
<linux/types.h> is included twice.

denisgabriel5 replied:
will remove it

dbogdan commented on drivers/iio/dac/ad5766.c:
```diff
@@ -6,35 +6,308 @@
+#define AD5766_CMD_SW_FULL_RESET		0x70
+#define AD5766_CMD_READBACK_REG(x)		(0x80 | ((x) & 0xF))
+#define AD5766_CMD_DITHER_SIG_1			0x90
+#define AD5766_CMD_DITHER_SIG_2			0xA0
```
Does your editor have the tabs set to 4 characters? They should be 8 characters - check https://github.com/analogdevicesinc/linux/blob/master/Documentation/process/coding-style.rst .

denisgabriel5 replied:
is set to 4 characters, will do a commit that will modify it

dbogdan commented on drivers/iio/dac/ad5766.c:
```diff
@@ -6,35 +6,308 @@
+	AD5766_CHANNEL(8, (_bits)),					\
+	AD5766_CHANNEL(9, (_bits)),					\
+	AD5766_CHANNEL(10, (_bits)),				\
+	AD5766_CHANNEL(11, (_bits)),				\
```
The AD5766/AD5767 are 16-channel, 16-bit/12-bit DACs - so the difference is the resolution, not the number of channels.

denisgabriel5 replied:
oh okay, i'll add them in the next commit

dbogdan commented on drivers/iio/dac/ad5766.c:
```diff
@@ -66,7 +341,7 @@ MODULE_DEVICE_TABLE(of, ad5766_dt_match);
 	{ "ad5766", ID_AD5766 },
 	{ "ad5767", ID_AD5767 },
-	{}
+	{},
```
Unrelated to this commit.

dbogdan commented on drivers/iio/dac/ad5766.c:
```diff
@@ -6,35 +6,311 @@
+	case IIO_CHAN_INFO_CALIBSCALE:
+		if (val >= 32 || val < -32)
+			return -EINVAL;
+		break;
```
What are you doing with IIO_CHAN_INFO_CALIBBIAS and IIO_CHAN_INFO_CALIBSCALE?

denisgabriel5 replied:
they are here before opening this PR, I figured during this PR that they aren't useful, but I thought  to leave them since removing them might me off-topic

dbogdan replied:
https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/dac/ad5766.c - this is the master version - am I missing something?

denisgabriel5 replied:
Well, they have been added before opening this PR, but after that master version was merged. To be honest, this PR was meant to help me with a few tips about the reading & writing part, hence the `[RFC/RFT]` from the title.

dbogdan commented on drivers/iio/dac/ad5766.c:
```diff
@@ -6,35 +6,311 @@
+	AD5766_CHANNEL(15, (_bits)),		\
+};
+
+#define DECLARE_AD5767_CHANNELS(_name, _bits)	\
```
Isn't DECLARE_AD5767_CHANNELS same as DECLARE_AD5766_CHANNELS?

denisgabriel5 replied:
it is, but I thought to leave them since removing them might me off-topic

dbogdan replied:
Aren't they added only now, through this commit?

denisgabriel5 replied:
oh, my bad, i thought they were added before this PR

denisgabriel5 replied:
I think I made a few commits and then I opened this PR and that's why it looks like they were just added.

denisgabriel5 replied:
Should I remove the 5767 version and just stick with the 5766 version for 5767 variant too?

dbogdan replied:
If they are the same, yes.

dbogdan commented on drivers/iio/dac/ad5766.c:
```diff
@@ -6,35 +6,311 @@
+#include <linux/sysfs.h>
+
 #include <linux/iio/iio.h>
+#include <linux/iio/sysfs.h>
```
I might be wrong, but in this version of the driver, only:
   #include <linux/module.h>
   #include <linux/iio/iio.h>
   #include <linux/spi/spi.h>
are required. I don't see a reason of including anything new in this commit.

denisgabriel5 replied:
In the beginning, I didn't know which one is needed and which is not, so, I included more just to be sure. I was pretty sure they are going to be filtered at some point.

commodo replied:
generally: always try to include the minimal number of headers required;

denisgabriel5 replied:
Anyway, I removed them and the build still works, so, I guess you were right :smile:

dbogdan commented on drivers/iio/dac/ad5766.c:
```diff
@@ -6,35 +6,311 @@
+ */
+
+struct ad5766_chip_info {
+	unsigned long			int_vref;
```
These parts don't have any internal reference - according to the datasheet, "the voltage at the VREF pin provides the reference voltage for the all DAC channels". Since you anyway don't use it, I would remove int_vref completely for now.

denisgabriel5 replied:
okay

commodo commented:
let's close this and continue with other PRs
